README

lua-httpd
---------

  This package contains a simple networking extension library for
 Lua 5.0 which is implemented and exported from a C shared library
 (.so).

  It may be installed system-wide and used by your Lua scripts with
 ease.



Installation
------------

  Compile and install the library with

	make

	make install  (as root).

  This will install the library into the directory /usr/lib/lua/5.0 
 along with a helper package into /usr/share/lua50.



Usage
-----

  Once the library has been installed you can use it from your Lua
 scripts easily.

  Please see the documentation included in docs/ for details.



Demonstration Code
------------------

  The package comes complete with several small examples inside the
 API documentation which you can find beneath the docs/ directory.


  More complete sample code is included:

  httpd.lua
  ---------

   A simple HTTP server written in Lua using the primitives.  This
  server understands virtual hosts, and will serve using your systems
  installed MIME types database.

   After you've installed the library you may execute the HTTPD by
  running "make test" - this will serve the content beneath ./vhosts
  to clients.

   To stop the server fetch:

		http://localhost:4444/finish

   The vhosts directory will contain a sub-directory for each virtual
  host you wish to serve, for example:

		./vhosts/
		./vhosts/my-host-name/
		./vhosts/my-host-name/htdocs  - The web root
		./vhosts/my-host-name/logs    - Where logfiles are saved


  client.lua
  ----------

   A very simple HTTP client written in Lua using the primitives.



License
-------

  This code is distributed under the terms of the GNU General
 Lesser Public License.



Steve
--
http://www.steve.org.uk/


debian/changelog
lua-httpd (0.6-1) stable; urgency=low

  * Minor update:
    - Compatability fixes for Windows/mingw by John Murga.

 -- Steve Kemp <steve@steve.org.uk>  Wed, 26 Jul 2006 10:40:42 +0000

lua-httpd (0.5-2) stable; urgency=low

  * Minor update:
    - Include linkes to lua-fs in the API documentation.
    - Fixed the package description to be 100% accurate.
    - Package description now links to the homepage.
    - Strip .cvsignore files from the package.

 -- Steve Kemp <steve@steve.org.uk>  Sun, 6 Nov 2005 20:12:18 +0000

lua-httpd (0.5-1) stable; urgency=low

  * New upstream release.
    - Added some simple filesystem primitives.
    - Install API documentation.
    - Install examples.

 -- Steve Kemp <steve@steve.org.uk>  Mon, 31 Oct 2005 17:21:05 +0000

lua-httpd (0.4-2) stable; urgency=low

  * Require lua50 to be installed.

 -- Steve Kemp <steve@steve.org.uk>  Sun, 30 Oct 2005 03:09:10 +0000

lua-httpd (0.4-1) stable; urgency=low

  * Initial version of Debian package.

 -- Steve Kemp <steve@steve.org.uk>  Sun, 30 Oct 2005 02:41:55 +0000


client.lua
--  -*-mode: C++; style: K&R; c-basic-offset: 4 ; -*- */

--
--  A simple HTTP client written in Lua, using the socket primitives
-- in 'libhttpd.so'.
--
--
-- $Id: client.lua,v 1.7 2005-10-30 02:07:03 steve Exp $


--
-- load the socket library
--
socket = require( "libhttpd" );
print( "Loaded the socket library, version:\n  " .. socket.version );


--
--  Connect to localhost.
--
sock = socket.connect( "localhost", 80 );

if ( sock == nil ) then
   error( "Socket failed to connect" );
end


print("Socket " .. sock );

--
--  Make a simple HTTP request.
--
socket.write( sock, "GET / HTTP/1.0\r\nConnection: close\r\n\r\n" );


--
-- Holders for a) a simple read request, and b) the total response.
--
line     = "";
response = "";


--
-- Loop reading until things fail.
--
repeat
   print( "Read - chunk" );
   len,line = socket.read( sock )
   response = response .. line;
until len <= 0


--
-- Finished.
--
socket.close( sock );


--
-- Output the HTTP headers + data.
--
print( response );


libhttpd.c
/* -*-mode: C++; style: K&R; c-basic-offset: 4 ; -*- */

/**
 *  libhttpd.c
 *   Simple network primitives for use by the Lua 5.0 scripting engine.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General  Public License (LGPL)
 *  as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 *
 *  Steve Kemp
 *  ---
 *  http://www.steve.org.uk/
 *  
 *  Further modification and porting to Win32 and Linux by John Murga 2006  
 *
 */


/*
 *  Each of the functions which passes, or returns, a socket merely
 * accessess them via Lua's "tonumber", or "fromnumber" routines.
 * 
 *  This is suboptimal, however in the absense of threads it is likely
 * to work out well in practice.
 *
 */

#include <stdio.h>
#include <stdlib.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

#define __USE_BSD 1
#include <string.h>

#ifndef mingw
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <netdb.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>


#include <unistd.h>

#ifdef mingw

#include <winsock.h>
typedef int socklen_t;
typedef SOCKET t_socket;
typedef t_socket *p_socket;

#define SOCKET_INVALID (INVALID_SOCKET) 

#endif

#define MYNAME		"libhttpd"
#define VERSION	        "$Id: libhttpd.c,v 1.19 2006-07-26 09:33:44 steve Exp $"



#include "lua.h"
#include "lauxlib.h"



/*
 * Find and return the version identifier from our CVS marker.
 * The memory returned is static.
 */
char * getVersion( )
{
    char *start  = NULL;
    char *end    = NULL;
    int length   = 0;
    char *prolog = "Release " RELEASE " CVS ID v";


    static char marker[128] = {'\0'};

    /*
     * If we have already calculated the revision information
     * then return it.
     */
    if ( strlen( marker ) > 0 )
    {
	return( marker );
    }


    start = strstr( VERSION, ",v " );
    if ( start == NULL )
	return NULL;

    /* Add on the ",v " text. */
    start += 3;
   
    /* Now find the next space - after the version marker */
    end = strstr( start, " " );
    if ( end == NULL )
	return NULL;


    /* The length of the version marker revision ID. */
    length  = end - start;

    /* Add in the prolog + revision information */
    strncpy( marker, prolog, strlen( prolog ) );
    strncat( marker, start, length );

    return( marker );
}


/**
 * Return an error string to the LUA script.
 * @param L The Lua intepretter object.
 * @param info An error string to return to the caller.  Pass NULL to use the return value of strerror.
 */
static int pusherror(lua_State *L, const char *info)
{
    lua_pushnil(L);

    if (info==NULL)
	lua_pushstring(L, strerror(errno));
    else
	lua_pushfstring(L, "%s: %s", info, strerror(errno));

    lua_pushnumber(L, errno);
    return 3;
}


/**
 * Bind a socket to a port, and return it.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pBind(lua_State *L)
{
    int sockfd;
    struct sockaddr_in serv_addr;
    int on   = 1;
    int port = 0;

    if (lua_isnumber(L, 1))
	port =lua_tonumber(L, 1);
    else
	return( pusherror(L, "bind(int) requires a port number" ) );


    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    if (sockfd < 0) 
        return( pusherror(L, "ERROR opening socket") );


    /* Enable address reuse */
    setsockopt( sockfd, SOL_SOCKET, SO_REUSEADDR, (char *) on, sizeof(on) );

    memset( &serv_addr, '\0', sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_addr.s_addr = INADDR_ANY;
    serv_addr.sin_port = htons(port);

    /* bind */
    if (bind(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) 
	return( pusherror(L, ("ERROR on binding") ));

    /* queue five connections. */
    listen(sockfd,5);

    /* Return socket */
    lua_pushnumber(L, sockfd);
    return( 1 );
}




/**
 * Connect a socket to a host/port, and return it.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pConnect(lua_State *L)
{
    struct sockaddr_in sa;
    struct hostent *hp;
    int ret,sockfd;
    const char *host;
    int port = 0;

    if (lua_isstring(L, 1))
	host = lua_tostring(L, 1);
    else
	return( pusherror(L, "connect(string,int) incorrect first argument" ) );
    if (lua_isnumber(L, 2))
	port = lua_tonumber(L, 2);
    else
	return( pusherror(L, "connect(string,int) incorrect second argument" ) );

    /* Get the host. */ 
    hp = gethostbyname(host);
    memcpy((char *)&sa.sin_addr, (char *)hp->h_addr, hp->h_length);
    sa.sin_family = hp->h_addrtype;
    sa.sin_port = htons(port);
    sockfd = socket(hp->h_addrtype, SOCK_STREAM, 0);

    /* connect */
    ret = connect(sockfd, (struct sockaddr *)&sa, sizeof(sa));
    if ( ret < 0 )
	return( pusherror(L, "Failed to connect" ) );

    /* Return socket */
    lua_pushnumber(L, sockfd);
    return( 1 );
}



/**
 * Accept a new connection upon a socket.  Return the new client.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pAccept(lua_State *L)
{
    int sockfd, newsockfd;
    socklen_t clilen;
    struct sockaddr_in  cli_addr;

    if (lua_isnumber(L, 1))
	sockfd =lua_tonumber(L, 1);
    else
	return( pusherror(L, "accept(int) requires listening socket." ) );
    
    /* accept() */
    clilen = sizeof(cli_addr);
    newsockfd = accept(sockfd, (struct sockaddr *) &cli_addr, &clilen);

    /* Return error on failure */
    if (newsockfd < 0) 
	return( pusherror(L,"ERROR on accept"));

    
    /*
     * Return the new socket and the connecting IP address.
     */
    lua_pushnumber(L,newsockfd);
    lua_pushstring(L,inet_ntoa(cli_addr.sin_addr));
    return 2;
}



/**
 * Read from a socket, return the data and the length read.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pRead(lua_State *L)
{
    int sockfd;
    int n = 0;

    /* Buffer we read() into */
    char buffer[4096];
    memset( buffer, '\0', sizeof(buffer));
    
    if (lua_isnumber(L, 1))
	sockfd =lua_tonumber(L, 1);
    else
	return( pusherror(L, "read(int)" ) );
    
    /* Do the read */
    n = recv(sockfd,buffer,sizeof(buffer)-1,0);

    if ( n == -1 )
	return( pusherror(L, "Problem reading from socket" ) );

    /* Return the data, and the length of that data */
    lua_pushnumber(L, n );
    lua_pushlstring(L, buffer, n );
    
    return( 2 );
}



/**
 * Write data to a socket.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pWrite(lua_State *L)
{
    int sockfd;
    const char *data = NULL;
    int length       = 0;
    int bytesSent    = 0 ;

    /* Get the number of arguments. */
    int n = lua_gettop(L);

    if ( ( n != 2 ) && ( n != 3 ) )
	return( pusherror(L, "write(int,string,[size])" ) );

    if (lua_isnumber(L, 1))
	sockfd =lua_tonumber(L, 1);
    else
	return( pusherror(L, "write(int,string,[size])" ) );

    /*
     * Supplied the string only
     */
    if ( n == 2 )
    {
	if (lua_isstring(L, 2))
	    data = lua_tostring(L, 2);
	else
	    return( pusherror(L, "write(int,string,[size])" ) );

	/*
	 * Get the string length from Lua.  This takes account of
	 * embedded NULLs.
	 */
	length = lua_strlen(L, 2 );
    }
    else if ( n == 3 )
    {
	if (lua_isstring(L, 2))
	    data = lua_tostring(L, 2);
	else
	    return( pusherror(L, "write(int,string,[size])" ) );

	/*
	 * Caller provided the length.
	 */
	if (lua_isnumber(L, 3))
	    length = lua_tonumber(L, 3);
	else
	    return( pusherror(L, "write(int,string,[size])" ) );
    }


    /*
     * Loop sending the data.
     */
    while (bytesSent < length)
    {
	/* Send some */
	int sent = send( sockfd, data + bytesSent, length - bytesSent,0);
	
	if (sent <= 0)
	{
	    return( pusherror(L, "Problem writing to socket" ) );
	}

	bytesSent += sent ;
    }   

    return 0;
}



/**
 * Close a socket.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pClose(lua_State *L)
{
    int sockfd;

    if (lua_isnumber(L, 1))
	sockfd =lua_tonumber(L, 1);
    else
	return( pusherror(L, "close(int)" ) );

    close( sockfd );

    return( 0 );
}



/**
 * Test to see if the given file/path is a directory.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pIsDir(lua_State *L)
{
    const char *fileName = NULL;
    struct stat st;

    if (lua_isstring(L, 1))
	fileName = lua_tostring(L, 1);
    else
	return( pusherror(L, "is_dir(string)" ) );

    if (stat(fileName, &st) == 0) 
	lua_pushboolean(L, S_ISDIR(st.st_mode) );
    else
	return( pusherror(L, "stat failed!" ) );

    return 1;
}



/**
 * Test to see if the given file/path is a file.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pIsFile(lua_State *L)
{
    const char *fileName = NULL;
    struct stat st;

    if (lua_isstring(L, 1))
	fileName = lua_tostring(L, 1);
    else
	return( pusherror(L, "is_file(string)" ) );

    if (stat(fileName, &st) == 0) 
	lua_pushboolean(L, S_ISREG(st.st_mode) );
    else
	return( pusherror(L, "stat failed!" ) );

    return 1;
}



/**
 * Return a table of all the directory entries in the given dir.
 * @param L The lua intepreter object.
 * @return The number of results to be passed back to the calling Lua script.
 */
static int pReadDir(lua_State *L)
{
    const char *dirName = NULL;
    struct dirent *dp;
    DIR *dir;
    int count = 0;
       
    if (lua_isstring(L, 1))
	dirName = lua_tostring(L, 1);
    else
	return( pusherror(L, "readdir(string)" ) );

    /* creates a table/array to hold the results */
    lua_newtable(L);

    dir = opendir(dirName);
    while (dir) 
    {
	if ((dp = readdir(dir)) != NULL) 
        {
	    /* Store in the table. */
	    lua_pushnumber(L, count);
	    lua_pushstring(L, dp->d_name );
	    lua_settable(L, -3);

	    count += 1;
	}
	else
	{
	    closedir( dir );
	    dir = 0;
	}
    }


    /* Make sure LUA knows how big our table is. */
    lua_pushliteral(L, "n");
    lua_pushnumber(L, count -1 );
    lua_rawset(L, -3);

    return 1;
}



/**
 * Mappings between the LUA code and our C code.
 */
static const luaL_reg R[] =
{
    {"bind",		pBind},
    {"connect",		pConnect},
    {"accept",		pAccept},
    {"close",           pClose},
    {"read",		pRead},
    {"write",		pWrite},
    {"is_dir",          pIsDir},
    {"is_file",         pIsFile},
    {"readdir",         pReadDir},
    {NULL,		NULL}
};



/**
 * Bind our exported functions to the Lua intepretter, making our functions
 * available to the calling script.
 * @param L The lua intepreter object.
 * @return 1 on success, 0 on failure.
 */
LUALIB_API int luaopen_libhttpd (lua_State *L)
{

#ifdef mingw

	WORD sockVersion;
	WSADATA wsaData;

	sockVersion = MAKEWORD(1, 1);
	WSAStartup(sockVersion, &wsaData);
#endif

    /* Version number from CVS marker. */
    char *version = getVersion();

    luaL_openlib(L, MYNAME, R, 0);
    lua_pushliteral(L,"version");
    lua_pushstring(L, version );
    lua_settable(L,-3);

    return 1;
}



default.lua
if not _REQUIREDNAME then
        error('Don\'t load directly, use require "libhttpd"', 2)
      end
      
      -- Get libname.
      local libname  = os.getenv "LUA_SOPATH"
      if libname then
        if not string.find(libname, '/$') then
          libname = libname .. "/"
        end
      else
        libname = "/usr/lib/lua/5.0/"
      end
      libname = libname .. _REQUIREDNAME .. ".so"

      
      local funcname = "luaopen_" .. _REQUIREDNAME
      
      local f, err = loadlib(libname, funcname)
      if not f then
        error("loadlib('" .. libname .. "', '" .. funcname .. "') failed: " .. err, 2)
      end
      
      local function ret(s, ...)
        if s then
          return unpack(arg)
        end
        error(_REQUIREDNAME .. "'s " .. funcname .. " failed: " .. arg[1], 3)
      end
      return ret(pcall(f))

httpd.lua
--  -*-mode: C++; style: K&R; c-basic-offset: 4 ; -*- */

--
--  A simple HTTP server written in Lua, using the socket primitives
-- in 'libhttpd.so'.
--
--  This server will handle multiple virtual hosts.  The only requirement
-- is that each virtual host must have the documents located beneath
-- a common tree.  Note that directory indexes are not (yet) supported.
--
--  For example to serve the three virtual hosts:
--    lappy
--    bob
--    localhost
--
--  You should have a directory layout such as:
--
--    ./vhosts/
--
--    ./vhosts/bob/
--    ./vhosts/bob/htdocs/
--    ./vhosts/bob/htdocs/index.html    [etc]
--    ./vhosts/bob/logs/
--
--    ./vhosts/lappy/
--    ./vhosts/lappy/htdocs/
--    ./vhosts/lappy/htdocs/index.html  [etc]
--    ./vhosts/lappy/logs/
--
--    ./vhosts/localhost/
--    ./vhosts/localhost/htdocs/
--    ./vhosts/localhost/htdocs/index.html  [etc]
--    ./vhosts/localhost/logs/
--
--
--  (If you wish to allow for fully qualified hosts simply add symbolic
-- links to suit your hostnames.)
--
--
--  If a request arrives for a virtual host which you are not serving
-- then it will be passed to the faux server "default".  Simply create
-- a symbolic link to the server you wish to be your default host:
--
--    cd ./vhosts
--    ln -s lappy default
--
--
--
--  The code is NOT secure
--  ----------------------
--
--  By that I mean that I won't promise any security.  I take the obvious
-- step of filtering '/../' from incoming requests and Hostnames, but there
-- may be other weaknesses.
--
--  Error messages are escaped by default, to prevent XSS attacks.
--
--  DOS attacks might be possible, although again the obvious cases
-- are covered.
--
--
-- Steve Kemp
-- --
-- http://www.steve.org.uk/
--
-- $Id: httpd.lua,v 1.31 2005-10-31 17:08:22 steve Exp $


--
-- load the socket library
--
socket = require( "libhttpd" );


print( "\n\nLoaded the socket library, version: \n  " .. socket.version );


--
--  A table of MIME types.  
--  These values will be loaded from /etc/mime.types if available, otherwise
-- a minimum set of defaults will be used.
--
mime = {};




--
--  Start a server upon the given port, using the given
-- root directory path.
--
--  The root path is designed to contain sub-directories for
-- each given virtual host.
--
function start_server( port, root )
    running = 1;

    --
    --  Bind a socket to the given port
    --
    local listener = socket.bind( port );

    if ( listener == nil ) then
       error("Error in bind()");
    end

    --
    --   Print some status messages.
    -- 
    print( "\nListening upon:" );
    print( "  http://localhost:" .. port .. "/" );
    print( "\nLoading virtual hosts from beneath: " );
    print( "  "  .. root .. "\n\n");


    --
    --  Loop accepting requests.
    --
    while ( running == 1 ) do
        processConnection( root, listener );
    end


    --
    -- Finished.
    --
    socket.close( listener );
end




--
--  Process a single incoming connection.
--
function processConnection( root, listener ) 
    --
    --  Accept a new connection
    --
    client,ip = socket.accept( listener );

    found   = 0;  -- Found the end of the HTTP headers?
    chunk   = 0;  -- Count of data read from client socket
    size    = 0;  -- Total size of incoming request.
    code    = 0;  -- Status code we send to the client
    request = ""; -- Request body read from client.


    --
    --  Read in a response from the client, terminating at the first
    -- '\r\n\r\n' line - this is the end of the HTTP header request.
    --
    --  Also break out of this loop if we read ten packets of data
    -- from the client but didn't manage to find a HTTP header end.
    -- This should help protect us from DOSes.
    --
    while ( ( found == 0 ) and ( chunk < 10 ) ) do
	length, data = socket.read(client);
        if ( length < 1 ) then
	    found = 1;
        end

	size    = size + length;
	request = request .. data;

	position,len = string.find( request, '\r\n\r\n' );
	if ( position ~= nil )  then
	    found = 1;
	end

        chunk = chunk + 1;
    end


    --
    --  OK We now have a complete HTTP request of 'size' bytes long
    -- stored in 'request'.
    --
    
    --
    -- Find the requested path.
    --
    _, _, method, path, major, minor  = string.find(request, "([A-Z]+) (.+) HTTP/(%d).(%d)");

    --
    -- We only handle GET requests.
    --
    if ( method ~= "GET" ) then
        error = "Method not implemented";

        if ( method == nil ) then
	    error = error .. ".";
        else
            error = error .. ": " .. urlEncode( method );
        end

        size = sendError( client, 501, error );
        socket.close( client );
        return size, "501";
    end


    --
    -- Decode the requested path.
    --
    path = urlDecode( path );


    --
    -- find the Virtual Host which we need for serving, and find the
    -- user agent and referer for logging purposes.
    --
    _, _, host    = string.find(request, "Host: ([^:\r\n]+)");
    _, _, agent   = string.find(request, "Agent: ([^\r\n]+)");
    _, _, referer = string.find(request, "Referer: ([^\r\n]+)");


    --
    -- If there was no "Host:" header then use "default".
    --
    if ( host == nil ) then
        host = "default";
    end


    --
    --  Is this a host that we're dealing with?
    --
    --  For this code to work we would ideally use the "os.stat" patch
    -- against Lua 5.0 which can be found at:
    --
    --     http://lua-users.org/wiki/PeterShook
    --
    --  Instead we use our own "is_dir" function inside the socket
    -- library.
    --
    info = socket.is_dir( root .. host );
    if ( not info ) then
       host = 'default';
    end

    --
    -- If the request was for '/finish' then terminate ourselves
    --
    if ( path == "/finish" ) then
        running = 0;

        socket.write( client, "HTTP/1.0 200 OK\r\n" );
        socket.write( client, "Server: lua-httpd " .. socket.version .. "\r\n" );
        socket.write( client, "Content-type: text/html\r\n" );
        socket.write( client, "Connection: close\r\n\r\n" );
        socket.write( client, "<html><head><title>Server Terminated</title></head>" );
        socket.write( client, "<body><h1>Server Terminated</h1><p>As per your request.</p></body></html>");
    else
        
        --
        -- Otherwise attempt to handle the connection.
        --
        size, code = handleRequest( root, host, path, client );
    end

    if ( agent == nil )   then agent   = "-" end;
    if ( referer == nil ) then referer = "-" end;
    if ( code == nill )   then code    = 0 ; end;

    --
    -- Log the access in something resembling the Apache common
    -- log format.
    -- 
    -- Note: The logging does not write the name of the virtual host.
    --
    logAccess( root, host, ip, path, code, size, agent, referer, major, minor );


    if ( running == 0 ) then
        print( "Terminating as per request from " .. ip );
    end

    --
    --  Close the client connection.
    --
    socket.close( client );
end


--
--  Attempt to serve the given path to the given client
--
function handleDirectory( client, path, request )
 
     a = socket.readdir( path );

    socket.write( client, "HTTP/1.0 200 OK\r\n" );
    socket.write( client, "Server: lua-httpd " .. socket.version .. "\r\n" );
    socket.write( client, "Content-type: text/html\r\n" );
    socket.write( client, "Connection: close\r\n\r\n" );

    msg = "";
    msg = msg .. "<html><head><title>Files in " .. request .. "</title></head>";
    msg = msg .. "<body><h2>Files in " .. request .. "</h2>\n";
    msg = msg .. "<ul>\n";

    for i=0,table.getn(a) do
         item = a[i];
         if ( socket.is_dir( path .. "/" .. item ) ) then
                item = item .. "/";
         end
         msg = msg .. "<li><a href=\"" .. item .. "\">" .. item .. "</a></li>\n";
    end

    msg = msg .. "</ul></body></html>\n";
    socket.write( client, msg );

    return string.len(msg) ;

end



--
--  Attempt to serve the given path to the given client
--
function handleRequest( root, host, path, client )
    --
    -- Local file
    --
    file = path;


    --
    --  File must be beneath the vhost root.
    --
    file = root .. host .. "/htdocs" .. file ;

    --
    --  Attempt to sanitize the input Virtual Host + requested path.
    --
    file = string.strip( file );


    --
    -- Add a trailing "index.html" to paths ending in / if such
    -- a file exists.
    --
    -- Otherwise if it is a directory then serve it.
    --
    if ( string.endsWith( file, "/" ) ) then  
        tmp = file .. "index.html";
        if ( fileExists( tmp ) ) then
           file = tmp;
        else
           if ( socket.is_dir( file ) ) then
               size = handleDirectory( client, file, path ) ;
               socket.close( client );
               return size, "200";
           end
        end
    end


 

    --
    -- Open the file and return an error if it fails.
    --    
    if ( fileExists( file ) == false ) then 
        size = sendError( client, 404,"File not found " .. urlEncode( path ) );
        socket.close( client );
        return size, "404";
    end;


    --
    -- Find the suffix to get the mime.type.
    --
    _, _, ext  = string.find( file, "\.([^\.]+)$" );
    if ( ext == nil ) then
       ext = "html";   -- HACK
    end
    
    type = mime[ext];
    if ( type == nil ) then type = 'text/plain' ; end;

    --
    -- Send out the header.
    --
    socket.write( client, "HTTP/1.0 200 OK\r\n" );
    socket.write( client, "Server: lua-httpd " .. socket.version .. "\r\n" );
    socket.write( client, "Content-type: " .. type  .. "\r\n" );
    socket.write( client, "Connection: close\r\n\r\n" );

    --
    -- Read the file, and then serve it.
    --
    f       = io.open( file, "rb" );
    size    = fileSize( f );
    local t = f:read("*all")
    socket.write( client, t );
    f:close();

    return size, "200" ;
end



--
--  Send the given error message to the client.
--  Return the length of data sent to the client so we know what to log.
--
function sendError( client, status, str )
    message = "HTTP/1.0 " .. status .. " OK\r\n" ;
    message = message .. "Server: lua-httpd " .. socket.version .. "\r\n";
    message = message .. "Content-type: text/html\r\n";
    message = message .. "Connection: close\r\n\r\n" ;
    message = message .. "<html><head><title>Error</title></head>" ;
    message = message .. "<body><h1>Error</h1" ;
    message = message .. "<p>" .. str .. "</p></body></html>" ;

    socket.write( client, message );
    return string.len(message) ;
end



--
--  Utility function:  Determine the size of an open file.
--
function fileSize (file)
    local current = file:seek()      -- get current position
    local size = file:seek("end")    -- get file size
    file:seek("set", current)        -- restore position
    return size
end


--
--  Utility function:  Determine whether the given file exists.
--
function fileExists (file)
    local f = io.open(file, "rb");
    if f == nil then
        return false;
    else
        f:close();
        return true;
    end
end


--
--  Read the mime file and setup mime types
--
function loadMimeFile(file, table)
     local f = io.open( file, "r" );
     while true do
         local line = f:read()
         if line == nil then break end
         _, _, type, name = string.find( line, "^(.*)\t+([^\t]+)$" );
         if ( type ~= nil ) then
             for w in string.gfind(name, "([^%s]+)") do
                  table[w] = type;
             end
         end
     end
end


--
--  Utility function:   Does the string end with the given suffix?
--
function string.endsWith(String,End)
    return End=='' or string.sub(String,-string.len(End))==End
end


--
--  Strip path traversal requests.
--
function string.strip( str )
    return( string.gsub( str, "/../", "" ) );
end


--
-- Utility function:  URL encoding function
--
function urlEncode(str)
    if (str) then
        str = string.gsub (str, "\n", "\r\n") 
        str = string.gsub (str, "([^%w ])",
            function (c) return string.format ("%%%02X", string.byte(c)) end) 
        str = string.gsub (str, " ", "+") 
    end 
    return str 
end


--
-- Utility function:  URL decode function
--
function urlDecode(str)
    str = string.gsub (str, "+", " ") 
    str = string.gsub (str, "%%(%x%x)", function(h) return string.char(tonumber(h,16)) end) 
    str = string.gsub (str, "\r\n", "\n") 
    return str 
end


--
-- Log an access request.
--
function logAccess( root, host, ip, request, status, size, agent, referer, major, minor )
    date = os.date("%m/%b/%Y:%H:%M:%S +0000");

    --
    -- Format the logging line.
    --
    log  = string.format( '%s - - [%s] "GET %s HTTP/%s.%s" %s %s "%s" "%s"', ip, date, request, major, minor, status, size, referer, agent );

    logfile = root ..  host .. "/logs/access.log"

    --
    -- Open logfile for appending.
    --
    file = io.open( logfile , "a" );
    if ( file ~= nil ) then
        file:write( log .. "\n" );
        file:close();
    else
        print( "WARNING : Unable to open logfile for writing : " .. logfile );
    end

    print( log );
end




--
-----
---
--
--  This is the start of the real code.  Now that our functions have been
-- defined we actually execute from this point onwards.
--
---
----
--



--
--  Setup the MIME types our server will use for serving files.
--
if ( fileExists( "/etc/mime.types" ) ) then
    loadMimeFile( "/etc/mime.types",  mime );
else 
    --
    --  The global MIME types file does not exist.
    --  Setup minimal defaults.
    --
    print( "WARNING: /etc/mime.types could not be read." );
    print( "         Running with minimal MIME types." );

    mime[ "html" ]  = "text/html";
    mime[ "txt"  ]  = "text/plain";
    mime[ "jpg"  ]  = "image/jpeg";
    mime[ "jpeg" ]  = "image/jpeg";
    mime[ "gif"  ]  = "image/gif";
    mime[ "png"  ]  = "image/png";
end


--
--  Now start the server.
--
--
start_server( 4444, "./vhosts/" );
